### PR TITLE
Finished structure polish

### PR DESCRIPTION
**Pull request by shunjizhan**
- added filecomponent and foldercomponent, and made then default prop for folderTree class

### CODE REVIEW

File: public/index.html [hunk 1]
```diff
@@ -3,8 +3,8 @@
   <head>
     <meta charset="utf-8">
     <meta name="viewport" content="width=device-width, initial-scale=1">
-    <link rel="shortcut icon" href="%PUBLIC_URL%/favicon.ico">
-    <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css">
+    <!-- <link rel="shortcut icon" href="%PUBLIC_URL%/favicon.ico"> -->
```
Comment (mboperator) [inline public/index.html]: We shouldn't commit commented code.

File: src/Components/FolderTree.jsx [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+    super(props);
+    this.setRootStatus = this.setRootStatus.bind(this);
+    this.state = {
+      data: initialize(props.data),
+    	checked: 0
+    };
+  }
+
+  setRootStatus(id, status) {
+  	// console.log('set rootStatus ', status)
```
Comment (mboperator) [inline src/Components/FolderTree.jsx]: Need to remove commented code

File: src/Components/FolderTree.jsx [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+	      	checked={this.state.data.status}
+
+          fileComponent={this.props.fileComponent}
+          folderComponent={this.props.folderComponent}
+	      />
+	    )
+	}
+}
+
+function filterAllSelected(node, rootFlag = false) {
```
Comment (mboperator) [inline src/Components/FolderTree.jsx]: We should add unit tests around these functions

File: public/index.html (hunk 1 above)
Comment (shunjizhan) [inline public/index.html]: DONE

File: src/Components/FolderTree.jsx (hunk 2 above)
Comment (shunjizhan) [inline src/Components/FolderTree.jsx]: DONE

File: src/Components/FileComponent.jsx [hunk 4]
```diff
@@ -0,0 +1,29 @@
+import React from 'react';
+import Checkbox from './Checkbox';
+import styles from './folderTreeCSS.css'
+
+const FileComponent = ({ level, checked, handleCheck, filename }) => (
+  <div className={styles.file}>
```
Comment (shunjizhan) [inline src/Components/FileComponent.jsx]: DONE

File: src/Components/FileComponent.jsx [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+    {getInden(level)}
+    <Checkbox status={checked} handleCheck={handleCheck} />
+    {'   '}<i className={styles.fileIcon} /> {filename}
```
Comment (shunjizhan) [inline src/Components/FileComponent.jsx]: DONE

File: src/Components/FolderComponent.jsx [hunk 6]
```diff
@@ -2,14 +2,6 @@ import React from 'react';
...
-  <div className={styles.file}>
-    {getInden(level)}
-    <Checkbox status={checked} handleCheck={handleCheck} />
-    {'   '}<i className={styles.fileIcon} /> {filename}
-  </div>
-);
-
 const FolderComponent = ({ level, checked, handleCheck, filename, toggleFolder, open }) => (
   <div className={styles.folder}>
     {getInden(level)}
```
Comment (shunjizhan) [inline src/Components/FolderComponent.jsx]: DONE

File: src/Components/FolderTree.jsx [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+  	let newData = this.state.data;
+  	newData.status = status;
+  	this.setState({data: newData});
+  }
+
+  printSelectedFileTree() {
+    // should exist better way to clone, right now data is const, this might be extra step. i.e. can filter this.state.data directly.
+  	let dataCopy = JSON.parse(JSON.stringify(this.state.data));			
+ 		let selectedTree = JSON.stringify(filterAllSelected(dataCopy, true));
+ 		console.log(selectedTree);
```
Comment (shunjizhan) [inline src/Components/FolderTree.jsx]: DONE

File: src/Components/FolderTree.jsx [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+
+          fileComponent={this.props.fileComponent}
+          folderComponent={this.props.folderComponent}
+	      />
+	    )
+	}
+}
+
+function filterAllSelected(node, rootFlag = false) {
+  if (rootFlag && !node.status) {                    // if it is root and is unchecked
```
Comment (shunjizhan) [inline src/Components/FolderTree.jsx]: DONE

File: src/Components/FolderTree.jsx [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+          folderComponent={this.props.folderComponent}
+	      />
+	    )
+	}
+}
+
+function filterAllSelected(node, rootFlag = false) {
+  if (rootFlag && !node.status) {                    // if it is root and is unchecked 
+    return {};
+  } else if (node.children != null && node.children.length > 0) {
```
Comment (shunjizhan) [inline src/Components/FolderTree.jsx]: DONE

File: src/Components/FolderTree.jsx [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+      node.children[i] = filterAllSelected(node.children[i]);
+    }
+    return filterNode(node);
+  } else {
+    return node;
+  }
+}
+
+function filterNode(node) {
+  let children = node.children;                            // current node doesn't change, only filter children
```
Comment (shunjizhan) [inline src/Components/FolderTree.jsx]: DONE

File: src/Components/Checkbox.jsx [hunk 11]
```diff
@@ -28,12 +23,16 @@ class Checkbox extends React.Component {
...
-		this.props.handleCheck(e);
-	}
-
 	render() {
-		return <input type="checkbox" onChange={this.handleCheck} checked={this.props.status !== 1? false : true} ref={box => this.checkBox = box} className={styles.checkBox}/>
+		return (
+			<input 
+				type="checkbox" 
+				onChange={this.props.handleCheck} 
+				checked={this.props.status !== 1? false : true}
```
Comment (mboperator) [inline src/Components/Checkbox.jsx]: This ternary can be moved into its on variable

File: src/Components/FolderComponent.jsx [hunk 12]
```diff
@@ -2,31 +2,27 @@ import React from 'react';
...
-    {getInden(level)}
-    <Checkbox status={checked} handleCheck={handleCheck} />
-    {'   '}<i className={styles.fileIcon} /> {filename}
-  </div>
-);
-
 const FolderComponent = ({ level, checked, handleCheck, filename, toggleFolder, open }) => (
-  <div className={styles.folder}>
-    {getInden(level)}
+  <div className={styles.folder} style={{'margin-left': getInden(level) }}>
```
Comment (mboperator) [inline src/Components/FolderComponent.jsx]: In React you want to do `style={{ marginLeft: getInden(level) }}`
Notice the camelCased key

File: src/Components/FolderComponent.jsx [hunk 13]
```diff
@@ -2,31 +2,27 @@ import React from 'react';
...
+
 function getInden(level) {
-  let iden = '', i = 0;
-  while (i < level) {
-    iden += ' ';
-    i++;
-  }
-  return iden;
+  const margin = 5 * level;
+  return margin + 'px';
```
Comment (mboperator) [inline src/Components/FolderComponent.jsx]: We should use `template strings` here: eg: `${margin}px`

File: src/Components/FolderTree.jsx [hunk 14]
```diff
@@ -0,0 +1,105 @@
...
+  }
+
+  setRootStatus(id, status) {
+  	let newData = this.state.data;
+  	newData.status = status;
+  	this.setState({data: newData});
+  }
+
+  printSelectedFileTree() {
+  	let dataDeepClone = JSON.parse(JSON.stringify(this.state.data));
```
Comment (mboperator) [inline src/Components/FolderTree.jsx]: This should be a const

File: src/Components/FolderTree.jsx [hunk 15]
```diff
@@ -0,0 +1,105 @@
...
+    super(props);
+    this.setRootStatus = this.setRootStatus.bind(this);
+    this.state = {
+      data: initialize(props.data),
+    	checked: 0
+    };
+  }
+
+  setRootStatus(id, status) {
+  	let newData = this.state.data;
```
Comment (mboperator) [inline src/Components/FolderTree.jsx]: This should be a const, and right now you're mutating the existing state.
To easily clone you can do `const newData = {...this.state.data}`

File: src/Components/FolderTree.jsx [hunk 16]
```diff
@@ -0,0 +1,105 @@
...
+  }
+
+ 	render() {
+ 			this.printSelectedFileTree();
+ 			return (
+	      <TreeNode
+	      	key={this.state.data.id}
+	      	category={this.state.data.category}
+	      	filename={this.state.data.filename}
+	      	children={this.state.data.children? this.state.data.children : []}
```
Comment (mboperator) [inline src/Components/FolderTree.jsx]: You can make this more concise by doing `this.state.data.children || []`

File: src/Components/FolderComponent.jsx (hunk 12 above)
Comment (shunjizhan) [inline src/Components/FolderComponent.jsx]: DONE

File: src/Components/FolderComponent.jsx (hunk 13 above)
Comment (shunjizhan) [inline src/Components/FolderComponent.jsx]: DONE

File: src/Components/Checkbox.jsx (hunk 11 above)
Comment (shunjizhan) [inline src/Components/Checkbox.jsx]: DONE

File: src/Components/FolderTree.jsx (hunk 15 above)
Comment (shunjizhan) [inline src/Components/FolderTree.jsx]: DONE

File: src/Components/FolderTree.jsx (hunk 14 above)
Comment (shunjizhan) [inline src/Components/FolderTree.jsx]: DONE

File: src/Components/FolderTree.jsx (hunk 16 above)
Comment (shunjizhan) [inline src/Components/FolderTree.jsx]: DONE

**Pull request merged**